app/api/utility.py

```diff
@@ -1,31 +1,106 @@
     return nodes_dict
 
 
+async def create_federation_node_index():
+    """
+    Creates an index of nodes for federation, which is a dict where the keys are the node URLs, and the values are the node names.
+    Fetches the names and URLs of public Neurobagel nodes from a remote directory file, and combines them with the user-defined local nodes.
+    """
+    local_nodes = parse_nodes_as_dict(LOCAL_NODES)
+
+    node_directory_url = "https://raw.githubusercontent.com/neurobagel/menu/main/node_directory/neurobagel_public_nodes.json"
+    node_directory_response = httpx.get(
+        url=node_directory_url,
+    )
+    if node_directory_response.is_success:
+        public_nodes = {
+            add_trailing_slash(node["ApiURL"]): node["NodeName"]
+            for node in node_directory_response.json()
+        }
+    else:
+        warnings.warn(
+            f"""
+            Unable to fetch directory of public Neurobagel nodes from {node_directory_url}.
+            The federation API will only register the nodes defined locally for this API: {local_nodes}.
+
+            Details of the response from the source:
+            Status code {node_directory_response.status_code}
+            {node_directory_response.reason_phrase}: {node_directory_response.text}
+            """
+        )
+        public_nodes = {}
+
+    # This step will remove any duplicate keys from the local and public node dicts, giving priority to the local nodes.
+    FEDERATION_NODES.update(
+        {
+            **public_nodes,
+            **local_nodes,
+        }
+    )
+
+
+def check_nodes_are_recognized(node_urls: list):
+    """
+    Check that all node URLs specified in the query exist in the node index for the API instance.
+    If not, raise an informative exception where the unrecognized node URLs are listed in alphabetical order.
+    """
+    unrecognized_nodes = sorted(
+        set(node_urls) - set(FEDERATION_NODES.keys())
+    )  # Resulting set is sorted alphabetically to make the error message deterministic
+    if unrecognized_nodes:
+        raise HTTPException(
+            status_code=422,
+            detail=f"Unrecognized Neurobagel node URL(s): {unrecognized_nodes}. "
+            f"The following nodes are available for federation: {list(FEDERATION_NODES.keys())}",
+        )
+
+
+def validate_query_node_url_list(node_urls: list) -> list:
+    """Format and validate node URLs passed as values to the query endpoint, including setting a default list of node URLs when none are provided."""
+    # Remove and ignore node URLs that are empty strings
+    node_urls = list(filter(None, node_urls))
+    if node_urls:
+        node_urls = [add_trailing_slash(node_url) for node_url in node_urls]
+        # Remove duplicates while preserving order
+        node_urls = list(dict.fromkeys(node_urls))
+        check_nodes_are_recognized(node_urls)
+    else:
+        node_urls = list(FEDERATION_NODES.keys())
```

Comment:
```suggestion
        # default to searching over all known nodes
        node_urls = list(FEDERATION_NODES.keys())
```
suggest adding comment here to remind ourselves that this is different from `async def create_federation_node_index()`